### PR TITLE
DoH dns lookup with cloudflare

### PR DESCRIPTION
**Pull request by SheetJSDev**
This `dns` fallback uses cloudflare DoH to implement a basic lookup

### CODE REVIEW

File: src/node-fallbacks/dns.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
+/*
+ * NOTE: THIS IS A TEMPORARY FALLBACK!!!!!
```
Comment (evanwashere) [inline src/node-fallbacks/dns.js:2]: this is browser polyfill

File: src/node-fallbacks/dns.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+ * 
+ * To support basic lookup, it uses DoH with CloudFlare's endpoint 1.1.1.1
+ */
+import { callbackify } from 'util';
+
+import * as consts from './dns_constants';
```
Comment (evanwashere) [inline src/node-fallbacks/dns.js:8]: constants should live in same file

File: src/node-fallbacks/dns.js [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+  async resolvePtr() { throw new Error(`unimplemented resolvePtr`); }
+  async resolveSoa() { throw new Error(`unimplemented resolveSoa`); }
+  async resolveSrv() { throw new Error(`unimplemented resolveSrv`); }
+  async resolveTxt() { throw new Error(`unimplemented resolveTxt`); }
+  async reverse() { throw new Error(`unimplemented reverse`); }
+  setLocalAddress(ipv4 = '0.0.0.0', ipv6 = '::0') { throw new Error(`unimplemented setLocalAddress`); }
+  setServers() { throw new Error(`unimplemented setServers`); }
+};
+
+const global_cb = new ResolverCB();
```
Comment (evanwashere) [inline src/node-fallbacks/dns.js:90]: this is a strange way to implement it

File: src/node-fallbacks/dns.js [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+  resolveNaptr: global_async.resolveNaptr.bind(global_async),
+  resolveNs: global_async.resolveNs.bind(global_async),
+  resolvePtr: global_async.resolvePtr.bind(global_async),
+  resolveSoa: global_async.resolveSoa.bind(global_async),
+  resolveSrv: global_async.resolveSrv.bind(global_async),
+  resolveTxt: global_async.resolveTxt.bind(global_async),
+  reverse: global_async.reverse.bind(global_async),
+  setServers: global_async.setServers.bind(global_async),
+  Resolver: ResolverAsync,
+};
```
Comment (evanwashere) [inline src/node-fallbacks/dns.js:65]: ^

File: src/node-fallbacks/dns.js [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+  async resolvePtr() { throw new Error(`unimplemented resolvePtr`); }
+  async resolveSoa() { throw new Error(`unimplemented resolveSoa`); }
+  async resolveSrv() { throw new Error(`unimplemented resolveSrv`); }
+  async resolveTxt() { throw new Error(`unimplemented resolveTxt`); }
+  async reverse() { throw new Error(`unimplemented reverse`); }
+  setLocalAddress(ipv4 = '0.0.0.0', ipv6 = '::0') { throw new Error(`unimplemented setLocalAddress`); }
+  setServers() { throw new Error(`unimplemented setServers`); }
+};
+
+const global_async = new ResolverAsync();
```
Comment (evanwashere) [inline src/node-fallbacks/dns.js:31]: ^

**Review by evanwashere — CHANGES_REQUESTED** [review]

**Comment by SheetJSDev** [comment]
The unexpected implementation structure stems from the actual DNS API https://nodejs.org/dist/latest-v18.x/docs/api/dns.html .

[Node v6](https://nodejs.org/docs/latest-v6.x/api/dns.html) only had the top-level methods like `dns.resolve`.  Node v8 added the `Resolver` class and v10 added the `promises` API

**Constants**

Constants are in a separate file because each constant is actually exported twice!

https://nodejs.org/dist/latest-v18.x/docs/api/dns.html#error-codes

> The dnsPromises API also exports the above error codes, e.g., dnsPromises.NODATA.

For example, the main `dns` module has to export the constant `CONNREFUSED` AND it must be an element in the `promises` export.  The separate script reduces line noise but it can be duplicated in the main script if that is preferable.

**Resolver Classes and Methods**

The `dns` module exposes a `resolve4` function AND the `Resolver` class has a `resolve4` method.  The main difference is that the former uses the default settings (tuned with `dns.setServers`) and the latter uses instance settings (tuned with the `Resolver#setServers` method).  The exported functions are neatly implemented using an internal instance of `Resolver` and exposing bound methods.

**Pull request closed without merging**